[IMP] sale,website_sale_*: use Owl product configurator in frontend

ltinel opened the pull request:
This change adapts the backend product configurator (which has been recently
migrated to Owl [1]) to make it usable in the frontend as well, and uses it in
the frontend.

This change also adds support for storable products in the product configurator
(e.g. it lets the user know when a product is out of stock).

This change migrates the "Add to cart" snippet to use the Owl product
configurator. Note that this will change the behavior of preexisting "Add to
cart" snippets: such snippets will now add the "first" variant to the cart
(without opening the configurator), even if the snippet was configured to let
the user choose the variant. This was deemed acceptable since migrating those
snippets could be extremely slow.

Lastly, previously, the product configurator supported optional products with
multiple parents. This allowed, for example, to have circularities in the
optional products. However, supporting circularities causes issues:
- If 2 products are parents of each other and are selected in the product
configurator, then they will stay selected, even if their parent products are
removed.
- If 2 products are parents of each other and are added to the cart, then the
corresponding SOLs won't be properly linked (as the SOLs need to be linked on
creation, and they can't be created at the exact same time).

[1] https://github.com/odoo/odoo/pull/106511

Enterprise PR: https://github.com/odoo/enterprise/pull/56469
Upgrade PR: https://github.com/odoo/upgrade/pull/5896

task-3572125

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/153779).

chevalierv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+                basic_product_information['strikethrough_price'] = strikethrough_price
+        return basic_product_information
+
+    def _get_ptav_price_extra(self, ptav, currency, so_date, product_or_template):
```
nitpicking
```suggestion
    def _get_ptav_price_extra(self, ptav, currency, so_date, product_or_template):
        """ Override of `sale` to ... """
```

chevalierv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+            return self._apply_taxes_to_price(price_extra, product_or_template, currency)
+        return price_extra
+
+    def _get_strikethrough_price(self, product_or_template, pricelist, currency, date, price):
```
Please add docstrings :angel:

chevalierv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+        return None
+
+    def _show_product(self, product_template, parent_combination):
+        if request.is_frontend:
+            return product_template._is_add_to_cart_possible(parent_combination)
+        return super()._show_product(product_template, parent_combination)
```
Why not call super if `request.is_frontend`?

ltinel replied:
Fixed :) This doesn't actually change anything since `_is_add_to_cart_possible` performs all the necessary checks but it's cleaner with `super`.

chevalierv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+
+    @staticmethod
+    def _populate_currency_and_pricelist(kwargs):
+        if request.is_frontend:
+            website = request.website
+            kwargs.update({
+                'currency_id': website.currency_id.id,
+                'pricelist_id': website.pricelist_id.id,
+            })
```
I don't believe `request.is_frontend` is needed here. I expect this method to be called from the website anyway. If you believe it's still required, please put the check outside the method for clarity.
```suggestion
          website = request.website
          kwargs.update({
              'currency_id': website.currency_id.id,
              'pricelist_id': website.pricelist_id.id,
```

ltinel replied:
Indeed, this method will only ever be called for frontend requests, so we can remove the check. FYI, my intention was to make it clear that we only "override" the currency and pricelist for requests coming from the frontend (but I'm fine with removing it).

chevalierv commented on addons/website_sale/static/src/js/product/product.js:
```diff
@@ -0,0 +1,26 @@
+        strikethrough_price: { type: Number, optional: true },
+        can_be_sold: { type: Boolean, optional: true },
+        category_name: { type: String, optional: true },
+        currency_name: { type: String, optional: true },
```
nitpicking: Don't hesitate to add comments for weird params
```suggestion
        currency_name: { type: String, optional: true },  // For tracking
```

chevalierv commented on addons/website_sale/static/src/js/product/product.xml:
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+    <t t-inherit="sale.Product" t-inherit-mode="extension">
+        <xpath
+            expr="//div[@name='sale_product_configurator_quantity_buttons']"
+            position="attributes">
```
nitpicking: I know it doesn't feel natural but I would prefer applying our convention 100% or keeping all on one line
```suggestion
        <xpath
            expr="//div[@name='sale_product_configurator_quantity_buttons']"
            position="attributes"
        >
```

ltinel replied:
Done here and elsewhere (just note that none of the product configurator components in `sale` follow this convention, so I didn't adjust those).

chevalierv commented on addons/website_sale/static/src/js/website_sale_product_configurator.js:
```diff
@@ -1,125 +1,133 @@
+                    wSaleUtils.updateCartNavBar(values);
+                    wSaleUtils.showCartNotification(this.call.bind(this), values.notification_info);
+                }
+                this._getCombinationInfo($.Event('click', { target: $('#add_to_cart') }));
```
I'm not quite convinced this is needed anymore. I guess it's to reload the page and get the up-to-date stock.

Could you either add a comment explaining why this line is needed or remove it?

chevalierv commented on addons/website_sale/static/src/js/website_sale_product_configurator.js:
```diff
@@ -1,125 +1,133 @@
+    },
 
-        return this.optionalProductsModal.opened();
+    // Hook to append additional props in overriding modules.
```
nitpicking: Could you switch to proper docstrings? :angel: 

```suggestion
    /**
     * Hook to append additional props in overriding modules.
     *
     * @return ...
     */
```

ltinel replied:
Obviously :) Don't know why I didn't do this initially.

chevalierv commented on addons/website_sale/static/src/snippets/s_add_to_cart/options.js:
```diff
@@ -127,22 +127,13 @@ options.registry.AddToCart = options.Class.extend({
-        }
-        buttonEl.dataset.productVariantId = productVariantId;
+        buttonEl.dataset.productTemplateId = this.$target[0].dataset.productTemplate;
+        // We're aware that getting the product id this way is too simplistic, but it mimics the
+        // previous logic. We'll fix this later on.
```
So we don't forget :p
```suggestion
        // TODO LOTI VCR We're aware that getting the product id this way is too simplistic, but it mimics the
        // previous logic. We'll fix this later on.
```

ltinel replied:
Done, although we should probably rewrite this whole file :sweat_smile:

chevalierv reviewed (changes requested):
Hello @ltinel 

Good job on this task! I've only a few nitpicking comments or small questions but otherwise, it's good to me :+1: 
Kudos for refactoring and improving the testing _suite_ :tada: 

PS: Next time, please do the renaming in an other commit (or PR) :beaver:

ltinel commented on addons/sale/controllers/product_configurator.py:
```diff
@@ -56,18 +57,22 @@ def get_product_configurator_values(
             )
         if not combination:
             combination = product_template._get_first_possible_combination()
+        currency = request.env['res.currency'].browse(currency_id)
+        pricelist = request.env['product.pricelist'].browse(pricelist_id)
+        so_date = datetime.fromisoformat(so_date)
```
Because these calls to `browse` were made for each product (main and optionals), even though the results were identical for all products. In theory, this didn't result in extra DB calls, but it still added avoidable duplication (and there was probably a negligible overhead due to cache lookups).

ltinel commented on addons/sale/controllers/product_configurator.py:
```diff
@@ -19,6 +19,7 @@ def get_product_configurator_values(
         pricelist_id=None,
         ptav_ids=None,
         only_main_product=False,
+        **kwargs,
```
Done here and elsewhere :)

ltinel commented on addons/sale/controllers/product_configurator.py:
```diff
@@ -5,10 +5,10 @@
 
-    @route('/sale_product_configurator/get_values', type='json', auth='user')
-    def get_product_configurator_values(
+    @route(route='/sale_product_configurator/get_values', type='json', auth='user')
```
If I'm not mistaken, RPCs are always sent via POST, so this will probably not work. I'll double-check to make sure.

The reason I added `methods=['POST'],` elsewhere is because exposing a GET endpoint for methods that modify data poses a security risk (people may try to forge requests to said endpoint).

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
     @route(
-        '/sale_product_configurator/show_advanced_configurator',
-        type='json', auth='public', methods=['POST'], website=True,
+        route='/website_sale_product_configurator/show_dialog',
```
What about `should_show_product_configurator`? It's a bit clearer IMO.

Also, do we want to do the same in the `sale` module? I.e. `/sale/product_configurator/...`, or is it less important there? Maybe we can simply remove the `sale` prefix altogether there ? I.e. `/product_configurator/...`.

chevalierv replied:
ok for `should_show_product_configurator` :+1: 

Yes we can do it in `sale` too! Let's keep the prefix for the moment as it represent the module in which the route exists.

chevalierv commented on addons/sale/static/src/js/tours/product_configurator_tour_utils.js:
Those new helpers will be a huge help for testing the configurator! You did a great job on this :tada:

chevalierv commented on addons/sale/static/src/js/tours/product_configurator_tour_utils.js:
```diff
@@ -0,0 +1,315 @@
+    };
+}
+
+function assertProductOutOfStock(productName) {
+    return {
+        content: `Assert that ${productName} is out of stock`,
+        trigger: `
+            ${productSelector(productName)}
+            td.o_sale_product_configurator_qty:contains("Out of stock")
+        `,
+        extra_trigger: `
+            ${productSelector(productName)}
+            td.o_sale_product_configurator_qty:not(:has(input[name="product_quantity"]))
+        `,
+        isCheck: true,
+    };
+}
+
+function assertOptionalProductOutOfStock(productName) {
+    return {
+        content: `Assert that ${productName} is out of stock`,
+        trigger: `
+            ${productSelector(productName)}
+            td.o_sale_product_configurator_price:contains("Out of stock")
+        `,
+        extra_trigger: `
+            ${productSelector(productName)}
+            td.o_sale_product_configurator_price:not(:has(button:contains("Add")))
+        `,
+        isCheck: true,
+    };
+}
+
```
Shouldn't we move the test utils to the module they are used? I mean, it feels weird to have utils to check `strike-through price` or `out of stock` in a module that doesn't implements those features.

ltinel replied:
Hmm, I agree with the idea, but I wonder whether we really want to split those helpers in multiple modules (especially since they're only for tests and don't add any dependencies)? I'm fine with either, so let me know if you have a preference.

Feyensv replied:
We might consider stock in the backend configurator in the future, but even if we do, we'd still allow to add the product to the cart, so I doubt those checks make sense outside of `website_sale` :thinking:

ltinel replied:
I moved the `website_sale` and `website_sale_stock` specific helpers to their corresponding modules :)

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+        return super().sale_product_configurator_update_combination(*args, **kwargs)
+
+    @route(
+        route='/website_sale_product_configurator/get_optional_products',
+        type='json',
+        auth='public',
+        website=True,
```
As I mentioned above, we want to allow both GET and POST for methods that only read data :)

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
-            product_id=product.id,
-            add_qty=add_qty,
+        :param int product_template_id: The product being checked, as a `product.template` id.
+        :param list(int) ptav_ids: The combination of the product, as a list of
+                                   `product.template.attribute.value` ids.
```
I'm fine with your suggestion if that's what we do elsewhere, but I noticed we also use `combination` for PTAV recordsets, which might be a bit confusing (because we reassign the variable from a list of ids to a recordset).

BTW, `sale_product_configurator_get_values` also uses `ptav_ids`, so I'll standardize that one as well.

chevalierv replied:
Ok I did not see it that way, maybe `ptav_ids` is better then update the docstrings to be clearer since it's not the combination of the product but the list of selected attribute ids

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+            ],
+            **kwargs
+        )
+        line_ids = {main_product['product_template_id']: values['line_id']}
```
Unfortunately, `values` is not a line record but a map, and I need to be able to call `write` on the line, so I believe I do need the `browse` (and it's more efficient to browse all lines at once).

I did notice an issue in this method however: by not setting `linked_line_id` during the creation of the optional product lines, they might not be correctly merged with existing lines. I'll push a fix.

chevalierv replied:
Okay, I though it was a record, too bad then :(

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+                line_ids[option['product_template_id']] = option_values['line_id']
+                parent_ids[option['product_template_id']] = option['parent_product_template_id']
 
+        for line in request.env['sale.order.line'].browse(line_ids.values()):
+            parent_product_template_id = parent_ids.get(line.product_template_id.id)
+            if parent_product_template_id:
+                linked_line_id = line_ids[parent_product_template_id]
+                if linked_line_id != line.linked_line_id.id:
+                    line.sudo().write({'linked_line_id': linked_line_id})
```
With my fix (see other comment related to browse), I set `linked_line_id` directly on creation. This requires the optional products to be provided in order (which should always be the case) so I added a note in the docstring.

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+    def _get_basic_product_information(
+        self, product_or_template, pricelist, combination, currency=None, date=None, **kwargs
+    ):
+        """ Override to append website data. """
```
It would be nice to be able to "append" a description and additional params to super's docstring without having to duplicate the whole thing :)

ltinel commented on addons/sale/static/src/js/product/product.js:
```diff
@@ -23,6 +23,7 @@ export class Product extends Component {
         exclusions: Object,
         parent_exclusions: Object,
         parent_product_tmpl_ids: { type: Array, element: Number, optional: true },
+        price_info: { type: String, optional: true },
```
Yes, that's what I did initially, but @Feyensv suggested otherwise [1]. TBH, I have no strong opinion regarding this. Both options are fine IMO.

[1] https://github.com/odoo/enterprise/pull/56469#discussion_r1537420316

ltinel commented on addons/sale/static/src/js/product_configurator_dialog/product_configurator_dialog.js:
```diff
@@ -65,6 +73,8 @@ export class ProductConfiguratorDialog extends Component {
             }
             this._checkExclusions(this.state.products[0]);
+            // Use the currency id retrieved from the server if none was provided in the props.
+            this.props.currencyId ??= currency_id;
```
Done. Just so you know, I also made a few additional changes to get rid of the `getCurrencyId` method :)

ltinel commented on addons/sale/static/src/js/product_template_attribute_line/product_template_attribute_line.xml:
```diff
@@ -25,7 +25,7 @@
     </t>
     <!-- Attributes value templates -->
-    <t t-name="saleProductConfigurator.ptav-select">
+    <t t-name="sale.ptav_select">
```
I'd probably do this later to avoid moving files around (knowing that there's another PR which will move these files to `product`).

ltinel commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,263 @@
+        # `combination` into account.
+        has_optional_products = bool(product_template.optional_product_ids.filtered(
+            lambda op: self._show_product(op, combination)
+        ))
```
Hmm, I'd need to investigate this. On one hand, I agree it would be a bit cleaner, but on the other hand, we'd need to do exactly the same in `_should_show_product` (which filters out optional products that shouldn't be shown in the configurator) and `get_single_product_variant` (which indicates whether a given product has optional products). Otherwise there may be inconsistencies in the configurator. But I'm not entirely sure these 2 methods should return the same results.

ltinel commented on addons/website_sale/static/src/snippets/s_dynamic_snippet_products/000.js:
```diff
@@ -145,7 +145,7 @@ const DynamicSnippetProductsCard = WebsiteSale.extend({
 
             const inputPT = document.createElement('input');
             inputPT.setAttribute('name', 'product_template_id');
-            inputPT.classList.add('product_template_id');  // how can this logic be so ugly??? T_T
```
@livaios this was only used to retrieve the product template directly inside `_handleAdd` right? (I updated _handleAdd to no longer rely on the `class` in this commit)

Feyensv replied:
You can say I'm the culprit behind this part of your commit, I assume my 'how can this logic be so ugly??' :D

ltinel commented:
@chevalierv FYI, I made a few additional changes unrelated to your comments:
- I adjusted the code in `website_sale_stock` following some breaking changes in `stock`,
- I now disable the dialog's footer buttons if a selected product is out of stock or unsellable (and I added tests).

chevalierv commented on addons/sale/static/src/js/product_configurator_dialog/product_configurator_dialog.js:
```diff
@@ -65,6 +76,8 @@ export class ProductConfiguratorDialog extends Component {
             }
             this._checkExclusions(this.state.products[0]);
+            // Use the currency id retrieved from the server if none was provided in the props.
+            this.currency.id ??= currency_id;
```
Do you store anything else than `id` in `this.env.currency`?

ltinel replied:
Nope, only the id, but I need the wrapper object so that this change gets propagated to the env.

chevalierv replied:
Oh smart, so no need for getters anymore

ltinel commented:
@chevalierv I added a second commit to stop supporting optional products with multiple parents. Would you mind having a look? :) Thanks!

chevalierv commented on addons/sale/static/src/js/product_configurator_dialog/product_configurator_dialog.js:
Don't forget to update the `checkExclusions` method :smiley_cat: 

You can see what was added to support recursive optional products in this commit: https://github.com/odoo-dev/odoo/commit/85342d72eeb401f396091dc62746f8c8b9a5748e

ltinel replied:
Done. I also simplified the `_addProduct` method a bit further (it was modifying a list while iterating over it, which is discouraged as it can yield unexpected results).

Feyensv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,323 @@
+        return should_show_product
+
+    @staticmethod
+    def _populate_currency_and_pricelist(kwargs):
+        website = request.website
+        kwargs.update({
+            'currency_id': website.currency_id.id,
+            'pricelist_id': website.pricelist_id.id,
+        })
```
Question on the fly, which date do we take for frontend requests ?
(since we cannot and shouldn't rely on the cart date)

Feyensv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,323 @@
+
+    @staticmethod
+    def _apply_taxes_to_price(price, product_or_template, currency):
+        product_taxes = product_or_template.sudo().taxes_id.filtered(
```
```suggestion
        product_taxes = product_or_template.sudo().taxes_id._filter_taxes_by_company(request.env.company)
```
new util to use (for subsidiary companies logic)

Feyensv commented on addons/website_sale/controllers/product_configurator.py:
```diff
@@ -1,147 +1,323 @@
+            date=date,
+            round=False,
+        )
+        list_price = self._apply_taxes_to_price(
+            product_or_template._price_compute('list_price', currency=currency)[
+                product_or_template.id
+            ],
+            product_or_template,
+            currency
+        )
```
Avoid computing the list_price if you don't use it please.

Feyensv reviewed (changes requested):
Tech pass, can't wait to merge this :partying_face:

Feyensv commented on addons/sale/static/src/js/tours/product_configurator_tour_utils.js:
```diff
@@ -0,0 +1,235 @@
+            td.o_sale_product_configurator_qty
+            span:contains("${priceInfo}")
+        `,
+        isCheck: true,
```
Yeah, one of the multiple PIPU PRs, I've lost count :D

Feyensv commented:
Can we have a little rebase before marking the PR as ready for review? :pray:

Feyensv commented on addons/website_sale/static/src/js/product/product.js:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @return {String} - The formatted strikethrough price.
+     */
+    get getFormattedStrikethroughPrice() {
```
get get?

ltinel replied:
:sweat_smile:

robodoo commented:
@ltinel @Feyensv linked pull request(s) odoo/upgrade#5896 not ready. Linked PRs are not staged until all of them are ready.
